assignments/burmanm/week-5.scm

```diff
@@ -0,0 +1,583 @@
+;;        (list 6 7)))
+
+;; $88 = (1 (4 (9 16) 25) (36 49))
+
+(define (square-tree-map tree)
+  (map (lambda (sub-tree)
+	 (if (pair? sub-tree)
+	     (square-tree-map sub-tree)
+	     (square sub-tree)))
+       tree))
+
+;; Test returns the same as previous
+
+;; Exercise 2.31
+
+(define (square-tree tree) (tree-map square tree))
+
+(define (tree-map f tree)
+  (map (lambda (sub-tree)
+	 (if (pair? sub-tree)
+	     (tree-map f sub-tree)
+	     (f sub-tree)))
+       tree)) 
+
+;; Exercise 2.32
+
+;; This one was tricky to understand.. needed a bit of Googling..
+
+(define (subsets s)
+  (if (null? s)
+      (list nil)
+      (let ((rest (subsets (cdr s))))
+	(append rest (map
+		      (lambda (x)
+			(cons (car s) x))
+		      rest)))))
+
+;; Test..
+
+(subsets '(1 2 3))
+;; $94 = (() (3) (2) (2 3) (1) (1 3) (1 2) (1 2 3))
+
+;; Exercise 2.33
+
+;; From SICP
+(define (accumulate op initial sequence)
+  (if (null? sequence)
+      initial
+      (op (car sequence)
+	  (accumulate op initial (cdr sequence)))))
+
+(define (map p sequence)
+  (accumulate (lambda (x y)
+		(cons (p x) y)) nil sequence))
+
+(define (append seq1 seq2)
+  (accumulate cons seq2 seq1))
+
+(define (length sequence)
+  (accumulate
+   (lambda (x y) (+ y 1)) 0 sequence))
+
+;; Test
+
+;; scheme@(guile-user) [37]> (map square (list 1 2 3 4 5 6))
+;; $101 = (1 4 9 16 25 36)
+
+;; scheme@(guile-user) [37]> (append (list 1 2 3) (list 4 5 6))
+;; $96 = (1 2 3 4 5 6)
+
+;; scheme@(guile-user) [37]> (length (list 1 2 3 4 5 6 7 8 9))
+;; $98 = 9
+
+;; Exercise 2.34
+
+;; result = result*x + poly[i]
+
+(define (horner-eval x coefficient-sequence)
+  (accumulate (lambda (this-coeff higher-terms)
+		(+ (* higher-terms x) this-coeff))
+	      0
+	      coefficient-sequence))
+
+;; Test
+
+(horner-eval 2 (list 1 3 0 5 0 1))
+
+;; scheme@(guile-user) [37]> (horner-eval 2 (list 1 3 0 5 0 1))
+;; $109 = 79
+
+
+;; Exercise 2.35
+
+;; From 2.2.2.. 
+;; (define (count-leaves x)
+;;   (cond ((null? x) 0)
+;; 	((not (pair? x)) 1)
+;; 	(else (+ (count-leaves (car x))
+;; 		 (count-leaves (cdr x))))))
+
+(define (count-leaves t)
+  (accumulate + 0
+	      (map (lambda (item)
+		     (if (pair? item)
+			 (count-leaves item)
+			 1))
+		   t)))
+
+;; Test
+(define x (cons (list 1 2) (list 3 4)))
+(count-leaves x)
+(count-leaves (list x x))
+
+;; scheme@(guile-user) [38]> (count-leaves x)
+;; $111 = 4
+;; scheme@(guile-user) [38]> (count-leaves (list x x))
+;; $112 = 8
+
+;; Exercise 2.36
+
+(define (accumulate-n op init seqs)
+  (if (null? (car seqs))
+      nil
+      (cons (accumulate op init (map (lambda (x) (car x)) seqs))
+	    (accumulate-n op init (map (lambda (x) (cdr x)) seqs)))))
+
+;; Test
+
+(define s (list (list 1 2 3) (list 4 5 6) (list 7 8 9) (list 10 11 12)))
+(accumulate-n + 0 s)
+
+;; scheme@(guile-user) [50]> (accumulate-n + 0 s)
+;; $120 = (22 26 30)
+
+;; Exercise 2.37
+
+;; From SICP
+(define (dot-product v w)
+  (accumulate + 0 (map * v w)))
```

Comment:
I totally missed that. So scheme map can take multiple list... I had to use accumulate-n